### PR TITLE
Visual Studio + CMake build

### PR DESCRIPTION
**Pull request by sergey-chernikov**
Added support for building on Visual Studio (2015 was tested)

### CODE REVIEW

File: src/base58.c [hunk 1]
```diff
@@ -44,8 +45,9 @@ int btc_base58_decode(void* bin, size_t* binszp, const char* b58)
     size_t binsz = *binszp;
     const unsigned char* b58u = (const void*)b58;
     unsigned char* binu = bin;
-    size_t outisz = (binsz + 3) / 4;
-    uint32_t outi[outisz];
+    const size_t outisz = (binsz + 3) / 4;
+    const size_t outilen = sizeof(uint32_t) * outisz;
+    uint32_t *outi = (uint32_t *)alloca(outilen);
```
Comment (jonasschnelli) [inline src/base58.c]: Is there a reason to heap allocate those buffers? Won't this compile on Visual Studio as it is?
`alloca` seems to be the wrong choice.
From the man pages:
```
    alloca() is machine and compiler dependent; its use is discouraged.
```

Heap allocation should always go over the memory mapper (use `btc_malloc()`, `btc_free()`).

File: src/commontools.c [hunk 2]
```diff
@@ -8,19 +8,24 @@
 #include <btc/bip32.h>
 #include <btc/ecc.h>
 #include <btc/ecc_key.h>
-#include <btc/net.h>
+#ifdef HAVE_CONFIG_H
+#  include <btc/net.h>
```
Comment (jonasschnelli) [inline src/commontools.c]: Why include this only when `HAVE_CONFIG_H` is defined?

**Comment by jonasschnelli** [comment]
Thanks for the pull request!
Concept ACK, though please don't switch to heap allocations and never use malloc directly, always use the `btc_malloc()` and `btc_free()` mapper.

File: src/base58.c (hunk 1 above)
Comment (sergey-chernikov) [inline src/base58.c]: Hi Jonas,
alloca() doesn't heap allocates - it allocates data on stack instead, and it appears no btc_alloca() in the code. It replaces VLA (supported only by gcc and C99 and discouraged by clang, for example). Visual Studio doesn't support VLAs, that's why I submitted the PR

Comment (jonasschnelli) [inline src/base58.c]: Thanks for the explanation. Today I learned!
What makes me a bit worry is the fact that `alloca()` is not a standard C library function while,.. although it looks like that VLAs are "optional" in C++11.

What about trying to get rid of VLAs by just using fixed sizes? Not sure if its possible though.

Comment (sergey-chernikov) [inline src/base58.c]: Then I'll replace alloca() to fixed-size arrays where possible and to btc_malloc()/btc_free() where not.

Comment (jonasschnelli) [inline src/base58.c]: Sounds good.

File: src/commontools.c (hunk 2 above)
Comment (sergey-chernikov) [inline src/commontools.c]: Fixed this also to check for WITH_NET == 1

File: src/secp256k1/contrib/lax_der_parsing.c [hunk 3]
```diff
@@ -5,7 +5,9 @@
  **********************************************************************/
 
 #include <string.h>
-#include <secp256k1.h>
+#ifdef HAVE_CONFIG_H
```
Comment (jonasschnelli) [inline src/secp256k1/contrib/lax_der_parsing.c]: I think you have to remove the `src/secp256k1` changes. It's a unchanged subtree (upstream is https://github.com/bitcoin-core/secp256k1).

**Comment by jonasschnelli** [comment]
Looks good. Have reviewed some parts.
Will review again once the libsecp256k1 changes are removed.

**Comment by sergey-chernikov** [comment]
I've fixed the includes of secp256k1.h (no need to duplicate them in .c-files, IMHO)

**Comment by jonasschnelli** [comment]
There is one VLA fix missing.
```
test/eckey_tests.c:71:13: error: variable length array used [-Werror,-Wvla]
    char str[size];
```

Also, you can't submit changes to the src/libsecp256k1 subtree here... you have to submit that patch upstream https://github.com/bitcoin-core/secp256k1. Otherwise this repository would have to maintain its own copy/fork of libsecp256k1.

**Comment by jonasschnelli** [comment]
Otherwise it looks good. All tests are passing... 
I think travis should also have a build configuration that uses gcc and `CFLAGS='-Werror=vla'`. But thats something we can do after this has been merged.

**Comment by sergey-chernikov** [comment]
I hope nothing else is missing now.

**Comment by jonasschnelli** [comment]
There is still a change in the secp256k1 subtree: https://github.com/libbtc/libbtc/pull/92/files#diff-83a8d5dea63b256de0d284df87e1e2dd

:-)

**Comment by sergey-chernikov** [comment]
Oops - should be fixed now

**Comment by droark** [comment]
@jonasschnelli - Any updates on getting this committed? Armory will be using this eventually, and Visual Studio needs to be supported. (I'm hoping for a Core-esque build system eventually, but unless somebody wants to help out with that in the near future....)

@sergey-chernikov - Can you rebase and fix the branch conflicts? I'd really like to see this one cross the finish line.

**Comment by jonasschnelli** [comment]
Thanks @droark for following up here.
I still like some feedback about VLAs versus this PRs approach.
Are there downsides? What are the technical differences between VLAs and heap alloc (probably the same for most/all compilers)?

Is Visual Studio really desirable? I know many cross platform projects work with MinGW for windows.
If so, how could we achieve Visual Studio compatibility in the CI/Travis?

**Comment by droark** [comment]
@jonasschnelli - Gotcha. Can't help with the VLAs. I haven't seriously built code on Windows in ages.

Regarding VS compatibility, it is desirable for Armory, for now at least. The Windows builds are built using VS. Personally, for various reasons, I don't like the setup. I spent an afternoon trying to get it set up and wanted to give up in frustration. (The lack of documentation regarding the process didn't help.) I agree that MinGW is a superior setup that allows for "one stop shopping." Alas, long story short, a push to move Armory to a Core-esque setup was never finished and can't be recycled. I really want to rewrite it and am finally comfortable enough with Autotools to do it myself. Alas, there are only so many hours in the day, and there are quite a few other tasks to complete before switching the build system. PRs welcome if anybody wants to help. ;)

Anyway, if this PR isn't finished, I'm sure we can write our own patch that will enable VS, hopefully as a temporary kludge on the way to a better build system.

**Comment by sergey-chernikov** [comment]
@jonasschnelli, @droark - hi gents!
On VLA and alloca difference: https://stackoverflow.com/questions/3488821/is-alloca-completely-replaceable, please also read notes here: http://man7.org/linux/man-pages/man3/alloca.3.html
Also, on some platforms usage of alloca is discouraged - so we decided with @jonasschnelli (some time earlier) to use dynamic memory allocation.
Anyway, I'll try to resolve conflicts with this PR soon.

**Comment by sergey-chernikov** [comment]
@jonasschnelli, please have a look - all tests seem to pass (clang OSX test fails due to configuration issues, I guess) and conflicts are resolved.

**Comment by droark** [comment]
@sergey-chernikov - What's the nature of the macOS failure? Perhaps I can take a look at it.

**Comment by droark** [comment]
@sergey-chernikov @jonasschnelli - Looks like valgrind stalled out on Travis. I don't know if that's common for this library.

**Comment by sergey-chernikov** [comment]
Hi @jonasschnelli!
Is there something in this PR that prevents merging it? It's now used not only by me, so the demand increases.

**Comment by csjones** [comment]
@sergey-chernikov I'm using CMake on macOS and these changes successfully create a static library target with Xcode but it does not find `gmp.h` despite gmp being installed with homebrew. Checking `GMP_INSTALL_DIR` returns nothing when the path should be `/usr/local/Cellar/gmp/6.1.2`.

**Comment by sergey-chernikov** [comment]
@csjones, sorry, don't have MacOS X - if anyone can provide fix for this issue, please do.

**Comment by droark** [comment]
I have a Mac. I can take a look. I'll try to do that in the next day or two.

**Comment by jonasschnelli** [comment]
Thanks @sergey-chernikov for keeping this alive.
I will have a look soon.
Does this PR require rebase since new files where added after the creation?

**Comment by sergey-chernikov** [comment]
@jonasschnelli, looks like no need to rebase. @droark, please confirm the patch builds successfully with Visual Studio now in BitcoinArmory.

**Comment by droark** [comment]
@sergey-chernikov - My apologies for the delay. Was juggling some other stuff. Just compiled Armory with the latest libbtc and with your patch. Everything compiled just fine with Autotools. libbtc has a few warnings but none of them seem to have been introduced by your patch.

**Comment by droark** [comment]
@sergey-chernikov - I just noticed that the PR is breaking various *NIX builds, at least on Travis. Any chance you can take a look? @goatpig has some changes he'd like to upstream, but he can't until this gets pulled.

@jonasschnelli - A few people have been using this privately for awhile now. Seems to work fine, although there are obviously some issues according to Travis. Is there anything else you'd like to see before this is accepted? I believe that Core now uses Appveyor to confirm Windows/MSVS builds. So, it should be possible to automate testing across the board.

**Comment by sergey-chernikov** [comment]
Only one test fails now, and I can't the place where it happens. I think it's the case for master, too

**Comment by xloem** [comment]
@sergey-chernikov jonasschnelli gave me merge rights in this repository and I'd like to address old PR's where reasonable.  Do you still use this PR?

**Comment by sergey-chernikov** [comment]
Yes, we're using it in our multi-platform projects (without an issue so far).

File: src/commontools.c [hunk 4]
```diff
@@ -85,7 +92,8 @@ btc_bool hd_gen_master(const btc_chainparams* chain, char* masterkeyhex, size_t
 {
     btc_hdnode node;
     uint8_t seed[32];
-    assert(btc_random_bytes(seed, 32, true));
+    const btc_bool rc = btc_random_bytes(seed, 32, true);
```
Comment (xloem) [inline src/commontools.c:95]: Note: this fix is also addressed in #142 which is a smaller set of changes.  Rebasing/merging could be helpful.

File: src/commontools.c [hunk 5]
```diff
@@ -100,17 +108,17 @@ btc_bool hd_print_node(const btc_chainparams* chain, const char* nodeser)
         return false;
 
     size_t strsize = 128;
-    char str[strsize];
+    char str[128];
```
Comment (xloem) [inline src/commontools.c:111]: Note: duplicate specification of constant 128 here.  My style if I merge might be to change uses of `strsize` to `sizeof(str)`.

File: src/commontools.c (hunk 4 above)
Comment (xloem) [reply src/commontools.c]: This PR was first so the change should be here.

Comment (sergey-chernikov) [reply src/commontools.c]: Sure, will do once #142 is merged

File: src/commontools.c (hunk 5 above)
Comment (sergey-chernikov) [reply src/commontools.c:111]: This is not a constant duplication - this is C89 conformance. The arrays in C can't have dimension specified by a variable (and VS produces a compiler error) - so it should be either a #define, or just a numeric.

File: src/commontools.c [hunk 6]
```diff
@@ -107,20 +108,18 @@ btc_bool hd_print_node(const btc_chainparams* chain, const char* nodeser)
     if (!btc_hdnode_deserialize(nodeser, chain, &node))
         return false;
 
-    size_t strsize = 128;
     char str[128];
+    size_t strsize = sizeof(str);
```
Comment (sergey-chernikov) [inline src/commontools.c:112]: This is a solution, too, ofc. Thanks for helping out here, @xloem!

Comment (xloem) [reply src/commontools.c:112]: I bump into a lot of bugs so sometimes I really notice places where somebody could mess stuff up by changing one thing and not another.  Thanks addressing these things so long ago.

**Pull request merged**